target/cheri-common/cheri-translate-utils.h

```diff
@@ -395,3 +588,1661 @@ static inline void gen_check_cond_branch_target(DisasContext *ctx,
+
+    gen_cap_get_base(ctx, rega, tempa);
+    gen_cap_get_base(ctx, regb, tempb);
+    tcg_gen_setcond_i64(TCG_COND_GEU, tempa, tempa, tempb);
+    tcg_gen_and_i64(result, result, tempa);
+
+    // Check perms (a.perms & ~b.perms) == 0
+    gen_cap_load_pesbt(ctx, rega, tempa);
+    gen_cap_load_pesbt(ctx, regb, tempb);
+
+    tcg_gen_andc_i64(tempa, tempa, tempb);
+    tcg_gen_movi_i64(tempb, CAP_CC(FIELD_HWPERMS_MASK64));
+    tcg_gen_and_i64(tempa, tempa, tempb);
+    tcg_gen_movi_i64(tempb, 0);
+    tcg_gen_setcond_i64(TCG_COND_EQ, tempa, tempa, tempb);
+    tcg_gen_and_i64(result, result, tempa);
+
+    // Finally that both caps have valid bounds
+    gen_cap_load_bounds_valid(ctx, rega, tempa);
+    gen_cap_load_bounds_valid(ctx, regb, tempb);
+    tcg_gen_and_i64(result, result, tempa);
+    tcg_gen_and_i64(result, result, tempb);
+
+    tcg_temp_free_i64(tempa);
+    tcg_temp_free_i64(tempb);
+
+    cheri_tcg_printf_verbose("ccd", "Get reg %d subset reg %d: %d\n", rega,
+                             regb, result);
+}
+
+static inline void gen_cap_is_subset_and_tag_eq(DisasContext *ctx, int rega,
+                                                int regb, TCGv_i64 result)
+{
+    gen_cap_is_subset(ctx, rega, regb, result);
+
+    TCGv_i64 tempa = tcg_temp_new_i64();
+    TCGv_i64 tempb = tcg_temp_new_i64();
+
+    tcg_gen_ld8u_i64(tempa, cpu_env,
+                     gp_register_offset(rega) +
+                         offsetof(cap_register_t, cr_tag));
+    tcg_gen_ld8u_i64(tempb, cpu_env,
+                     gp_register_offset(regb) +
+                         offsetof(cap_register_t, cr_tag));
+
+    tcg_gen_xor_i64(tempa, tempa, tempb);
+    tcg_gen_andc_i64(result, result, tempa);
+
+    tcg_temp_free_i64(tempa);
+    tcg_temp_free_i64(tempb);
+}
+#endif
+
+static inline void gen_cap_set_pesbt(DisasContext *ctx, int regnum, TCGv pesbt)
+{
+    if (regnum == NULL_CAPREG_INDEX)
+        return;
+
+    gen_lazy_cap_set_state(ctx, regnum, CREG_UNTAGGED_CAP);
+    // Again, once I bother to change where we apply this mask this should go
+    // away
+    TCGv mem_pesbt = tcg_const_tl(CAP_NULL_XOR_MASK);
+    tcg_gen_xor_tl(mem_pesbt, mem_pesbt, pesbt);
+    tcg_gen_st_tl(mem_pesbt, cpu_env,
+                  gp_register_offset(regnum) +
+                      offsetof(cap_register_t, cr_pesbt));
+    tcg_temp_free(mem_pesbt);
+}
+
+// Checks a cap is in bounds for a given size, has specified perms, is tagged,
+// and not sealed. Or throws an exception. Because of the branch, this will kill
+// every temp. Addr is specially conserved as it will probably be used again.
+static inline void gen_cap_memop_checks(DisasContext *ctx, int regnum,
+                                        TCGv addr, target_ulong size, int perms)
+{
+#ifdef DO_TCG_BOUNDS_CHECKS
+    // We use addr as a tmp, and tmp as address because we can ensure tmp is
+    // local
+    TCGv_i64 local_addr = tcg_temp_local_new_i64();
```

Comment:
These should all be tl given they're not under AARCH64 defines (only happen to be due to that being the only TCG platform); (and yes, gen_cap_in_bounds expects result to be a tl).